### DISCUSSION TITLE
Replace generic parameters by typed options

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#7979] [Proposals]
**Describe the problem you would like to solve**
Generic parameters, passed through arrays are complex to deal with through a maintenance perspective, one also will have to dig into the documentation/code to see what are the possibilities/requirements (e.g. one argument depends on another).

**Describe the solution you'd like**
Given the PHP requirement was bumped to 8.0, perhaps we can adopt something like the code below. Which isn't very verbose and would allows us to:
- Add types to the parameters
- Force the input/presence of parameters
- Default values
- Increase the discoverability of them through the code editor
- Apply generic validations at the object creation

```php
class QueryOptions {
    public function __construct(
        public ?array $includeUserIds = null,
        public bool $includeInactive = false,
        public ?string $query = null)
    {
        // Validations
    }
}

new QueryOptions(query: 'search string');
```

**Who is asking for this feature?**
Developer

**Comment by NateWr** [comment]
Do you have some examples of code that could use this? A big part of the refactor from Service/QueryBuilder to Repo/Collector was to implement typed method parameters for exactly this reason. I'd prefer if this issue identified one or more sub-systems to be refactored, rather than just a generic recommendation. If it's a generic recommendation, it's probably something we should add to a [contributor's guide](https://github.com/pkp/pkp-docs/issues/732).

**Comment by jonasraoni** [comment]
Excluding the "grids" (which we're probably not going to touch), I've seen some cases here and there.
About sub-systems, besides the services/query builders, I could add the hooks and the form components.

**Comment by NateWr** [comment]
Thanks, the hooks sound like a promising use case. Can you put some brief code samples here to show how they might work with typed options?

**Comment by tib-rijo** [comment]
Related via the "parse, don't validate" pattern:

- https://github.com/pkp/pkp-lib/issues/10285

**Comment by asmecher** [comment]
We're well underway in adopting these conventions; closing the discussion.
